src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/CalculationTransactionalService.kt

```diff
@@ -39,18 +42,117 @@ class CalculationTransactionalService(
   private val prisonService: PrisonService,
   private val domainEventPublisher: DomainEventPublisher,
   private val prisonApiDataMapper: PrisonApiDataMapper,
-  private val calculationService: CalculationService
+  private val calculationService: CalculationService,
+  private val bookingService: BookingService,
+  private val validationService: ValidationService,
 ) {
 
   fun getCurrentAuthentication(): AuthAwareAuthenticationToken =
     SecurityContextHolder.getContext().authentication as AuthAwareAuthenticationToken?
       ?: throw IllegalStateException("User is not authenticated")
 
+  /*
+   * There are 4 stages of full validation:
+   * 1. validate user input data and the raw sentence and offence data from NOMIS
+   * 2. Validate the transformation of the raw Booking (NOMIS offence data transformed into a Booking object)
+   * 3. Run the calculation and catch any errors thrown by the calculation algorithm
+   * 4. Validate the post calculation Booking (The Booking is transformed during the calculation). e.g. Consecutive sentences (aggregates)
+   */
+  fun fullValidation(prisonerId: String, calculationUserInputs: CalculationUserInputs): List<ValidationMessage> {
+    val sourceData = prisonService.getPrisonApiSourceData(prisonerId)
+    var messages = validationService.validateSourceData(sourceData) // Validation stage 1 of 4
+    if (messages.isNotEmpty()) return messages
+
+    val booking = bookingService.getBooking(sourceData, calculationUserInputs)
+    messages = validationService.validateBookingBeforeCalculation(booking) // Validation stage 2 of 4
+    if (messages.isNotEmpty()) return messages
+    try {
+      val bookingAfterCalculation = calculationService.calculate(booking) // Validation stage 3 of 4
+      validationService.validateBookingAfterCalculation(bookingAfterCalculation) // Validation stage 4 of 4
+    } catch (validationException: CrdCalculationValidationException) {
+      return listOf(ValidationMessage(validationException.validation))
+    }
+    return messages
+  }
+
+  //  The activeDataOnly flag is only used by a test endpoint (1000 calcs test, which is used to test historic data)
+  @Transactional
+  fun calculate(prisonerId: String, calculationUserInputs: CalculationUserInputs, activeDataOnly: Boolean = true): CalculatedReleaseDates {
+    val sourceData = prisonService.getPrisonApiSourceData(prisonerId, activeDataOnly)
```

Comment:
Does this no longer run validate? Also its being used by for a test calculation (1000 calcs) and preliminary calcs, but always persisting PRELIMNARY as the calculation type.

Reply:
Good spot - changed

Reply:
as discussed, validation is used by 1000 calcs so reinsert

Reply:
Have re-inserted validation 1000 calcs and did an e2e test where it fails , it workedi.e. passed the errors to the 1000 calcs.
TODO (next PR) write a test for this